sklearn/cluster/k_means_.py

```diff
@@ -834,7 +836,11 @@ def __init__(self, n_clusters=8, init='k-means++', n_init=10,
 
     def _check_fit_data(self, X):
         """Verify that the number of samples given is larger than k"""
-        X = check_array(X, accept_sparse='csr', dtype=np.float64)
+        # to handle sparse data which only works as float64 at the moment
+        if sp.issparse(X):
+            X = check_array(X, accept_sparse='csr', dtype=np.float64)
+        else:
+            X = check_array(X, dtype=None)
```

Comment:
I think you should do:

```python
X = check_array(X, dtype=[np.float64, np.float32])
```

make it possible to pass integer values to fit and have them automatically converted to `np.float64`. Please add a test for this case as well.